experimental/SetPartitions/src/SetPartition.jl

```diff
@@ -0,0 +1,293 @@
+
+# Examples
+```jldoctest
+julia> vertical_reflection(SetPartition([1, 2, 3], [2, 1]))
+SetPartition([1, 2, 3], [3, 2])
+```
+"""
+function vertical_reflection(p::SetPartition)
+    return SetPartition(reverse(p.upper_points), reverse(p.lower_points))
+end
+
+"""
+    rotation(p::SetPartition)
+
+Return the rotation of `p` in the direction given by `lr` and `tb`. 
+
+Rotating a partition moves the left- or right-most point of the upper points 
+to the lower points or vice verca. See also Section 4.1.2 in [Gro20](@cite).
+
+# Arguments
+- `p`: input partition
+- `lr`: rotating at the left (true) or at the right (false)
+- `tb`: rotating from top to bottom (true) or from bottom to top (false)
+
+# Examples
+```jldoctest
+julia> rotation(SetPartition([1, 2, 3], [2, 1]), true, true)
+SetPartition([1, 2], [3, 1, 3])
+
+julia> rotation(SetPartition([1, 2, 3], [2, 1]), true, false)
+SetPartition([1, 2, 1, 3], [2])
+
+julia> rotation(SetPartition([1, 2, 3], [2, 1]), false, true)
+SetPartition([1, 2], [2, 1, 3])
+
+julia> rotation(SetPartition([1, 2, 3], [2, 1]), false, false)
+SetPartition([1, 2, 3, 1], [2])
+```
+"""
+function rotation(p::SetPartition, lr::Bool, tb::Bool)
+    
+    if tb && isempty(p.upper_points)
+        error("SetPartition has no top part")
+    elseif !tb && isempty(p.lower_points)
+        error("SetPartition has no bottom part")
+    end
+
+    ret = (copy(p.upper_points), copy(p.lower_points))
+
+    if lr
+        if tb
+            a = ret[1][1]
+            splice!(ret[1], 1)
+            pushfirst!(ret[2], a)
+        else
+            a = ret[2][1]
+            splice!(ret[2], 1)
+            pushfirst!(ret[1], a)
+        end
+    else
+        if tb
+            a = ret[1][end]
+            pop!(ret[1])
+            push!(ret[2], a)
+        else
+            a = ret[2][end]
+            pop!(ret[2])
+            push!(ret[1], a)
+        end
+    end
+    
+    return SetPartition(ret[1], ret[2])
+end
+
+"""
+    composition_loops(p::SetPartition, q::SetPartition)
+
+Return the composition of `p` and `q` as well as the number of removed loops.
+
+The composition of two partitions is obtained concatenating them horizontally
+and removing intermediate loops which are no longer connected to the top or bottom.
+See also Section 4.1.1 in [Gro20](@cite).
+
+The composition of `p` and `q` is only defined if the number of upper points of 
+`p` equals the number of lower points of `q`. See also [`is_composable`](@ref).
+
+# Examples
+```jldoctest
+julia> composition_loops(SetPartition([1, 2], [2, 1]), SetPartition([1], [1, 1]))
+(SetPartition([1], [1, 1]), 0)
+
+julia> composition_loops(SetPartition([1, 1], [2]), SetPartition([1], [2, 2]))
+(SetPartition([1], [2]), 1)
+
+julia> composition_loops(SetPartition([1], [1, 2]), SetPartition([1], [2, 2]))
+ERROR: format not fitting
+```
+"""
+function composition_loops(p::SetPartition, q::SetPartition)
+    
+    !is_composable(p, q) ? error("format not fitting") : 
```

Comment:
```suggestion
    @req is_composable(p, q) "Number of points mismatch" 
```
or similar. Please use `@req` instead of `error(...)` for input sanitation

Reply:
Hi, thanks for your feedback! This @req macro does not seem to be part of the julia standard library. Do you mean by using @req that I should implement my own req macro or should I use the @req from DiscreteValueIteration (https://github.com/JuliaPOMDP/DiscreteValueIteration.jl) or am I missing sth?

Reply:
This is part of Hecke and thus available in Oscar without any extra work. See https://docs.oscar-system.org/dev/Hecke/features/macros/#Miscellaneous for the docs.

Reply:
But I agree that this should be added to the Oscar developer docs

Reply:
ah I see, makes sense, thanks!